stl/src/tzdb.cpp

```diff
@@ -1,15 +1,302 @@
 // Copyright (c) Microsoft Corporation.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
+#include <atomic>
+#include <icu.h>
 #include <internal_shared.h>
+#include <memory>
+#include <string_view>
+#include <xfilesystem_abi.h>
 #include <xtzdb.h>
 
 #include <Windows.h>
 
 #pragma comment(lib, "Advapi32")
 
+namespace {
+    enum class _Icu_api_level : unsigned long {
+        _Not_set,
+        _Detecting,
+        _Has_failed,
+        _Has_icu_addresses,
+    };
+
+    struct _Icu_functions_table {
+        _STD atomic<decltype(&::ucal_getCanonicalTimeZoneID)> _Pfn_ucal_getCanonicalTimeZoneID{nullptr};
+        _STD atomic<decltype(&::ucal_getDefaultTimeZone)> _Pfn_ucal_getDefaultTimeZone{nullptr};
+        _STD atomic<decltype(&::ucal_openTimeZoneIDEnumeration)> _Pfn_ucal_openTimeZoneIDEnumeration{nullptr};
+        _STD atomic<decltype(&::uenum_close)> _Pfn_uenum_close{nullptr};
+        _STD atomic<decltype(&::uenum_count)> _Pfn_uenum_count{nullptr};
+        _STD atomic<decltype(&::uenum_unext)> _Pfn_uenum_unext{nullptr};
+        _STD atomic<_Icu_api_level> _Api_level{_Icu_api_level::_Not_set};
+    };
+
+    _Icu_functions_table _Icu_functions;
+
+    template <typename T>
+    void _Load_address(const HMODULE _Module, _STD atomic<T>& _Stored_Pfn, LPCSTR _Fn_name, DWORD& _Last_error) {
+        const auto _Pfn = reinterpret_cast<T>(GetProcAddress(_Module, _Fn_name));
+        if (_Pfn != nullptr) {
+            _Stored_Pfn.store(_Pfn, _STD memory_order_relaxed);
+        } else {
+            _Last_error = GetLastError();
+        }
+    }

```

Comment:
I did this because the code repetition was getting out of control, particular as I have more than double the amount of functions in the next PR. I considered following atomic but it felt like overkill particularly as no tzdb timezone function should run without _Icu_functions._Api_level set correctly. 

`atomic_wait.cpp` seems to have backup functionality if functions aren't loaded so having them all nulled out might be more important there.

Reply:
That makes sense to me!  In the case of the backup functionality in `atomic_wait.cpp`, we can always revisit this scheme and reconsider whether we need to null things out when we come back to implement a fallback for before-19H1 OSes.  I'll add a note to the tracking issue to consider this when we get there.